File: src/com/easy/DegreeOfAnArray_697.java
/**
 * 
 */
package com.easy;

import java.util.Collections;
import java.util.HashMap;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class DegreeOfAnArray_697 {
    // 2019年8月14日
    // Use hash map
    // Time Complexity: O(n)
    // Space Complexity:O(n)
    // Runtime: 29 ms, faster than 35.00%
    // Memory Usage: 41.1 MB, less than 88.89%
    public int findShortestSubArray(int[] nums) {
        HashMap<Integer, Integer> mMapCount = new HashMap<>();
        HashMap<Integer, Integer> mMapLeft = new HashMap<>();
        HashMap<Integer, Integer> mMapRight = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            mMapCount.put(nums[i], mMapCount.getOrDefault(nums[i], 0) + 1);
            if (!mMapLeft.containsKey(nums[i])) {
                mMapLeft.put(nums[i], i);
            }
            mMapRight.put(nums[i], i);
        }
        int mResult = nums.length;
        int mDegree = Collections.max(mMapCount.values());
        System.out.println(mDegree);
        for (int key : mMapCount.keySet()) {
            if (mMapCount.get(key) == mDegree) {
                mResult = Math.min(mResult, mMapRight.get(key) - mMapLeft.get(key) + 1);
            }
        }

        return mResult;
    }
}

File: Golang_src/easy/degreeOfAnArray_697.go
package easy

import "math"

// Use hashmap:
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 34 ms, faster than 30.77%
// Memory Usage: 7.1 MB, less than 23.8%
func findShortestSubArray(nums []int) int {
	result := math.MaxInt32

	max := func(a, b int) int {
		if a > b {
			return a
		}
		return b
	}
	min := func(a, b int) int {
		if a < b {
			return a
		}
		return b
	}

	maxDeg := 0

	// [1,2,2,3,1]
	// 1: [0,4]
	// 2: [1,2]
	// 3: [3]
	hashmap := map[int][]int{}
	for i, v := range nums {
		if val, ok := hashmap[v]; ok {
			val = append(val, i)
			hashmap[v] = val
			maxDeg = max(maxDeg, len(val))
		} else {
			hashmap[v] = []int{i}
			maxDeg = max(maxDeg, 1)
		}
	}

	for _, v := range hashmap {
		if len(v) == maxDeg {
			result = min(result, v[len(v)-1]-v[0]+1)
		}
	}

	return result
}
